qiskit/algorithms/minimum_eigensolvers/adapt_vqe.py

```diff
@@ -0,0 +1,280 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""An implementation of the AdaptVQE algorithm."""
+
+from __future__ import annotations
+
+from enum import Enum
+from typing import Optional, Sequence
+
+import copy
+import re
+import logging
+
+import numpy as np
+
+from qiskit import QiskitError
+from qiskit.algorithms.list_or_dict import ListOrDict
+from qiskit.opflow import OperatorBase, PauliSumOp, commutator
+from qiskit.circuit.library import EvolvedOperatorAnsatz
+from qiskit.utils.validation import validate_min
+
+from .vqe import VQE, VQEResult
+from ..observables_evaluator import estimate_observables
+from ..variational_algorithm import VariationalAlgorithm
+
+
+logger = logging.getLogger(__name__)
+
+
+class TerminationCriterion(Enum):
+    """A class enumerating the various finishing criteria."""
+
+    CONVERGED = "Threshold converged"
+    CYCLICITY = "Aborted due to a cyclic selection of evolution operators"
+    MAXIMUM = "Maximum number of iterations reached"
+
+
+class AdaptVQE(VariationalAlgorithm):
+    """The Adaptive Variational Quantum Eigensolver algorithm.
+
+    `AdaptVQE <https://arxiv.org/abs/1812.11173>`__ is a quantum algorithm which creates a compact
+    ansatz from a set of evolution operators. It iteratively extends the ansatz circuit, by
+    selecting the building block that leads to the largest gradient from a set of candidates. In
+    chemistry, this is usually a list of orbital excitations. This results in a wavefunction ansatz
+    which is uniquely adapted to the operator whose minimum eigenvalue is being determined.
+    This class relies on a supplied instance of :class:`~.VQE` to find the minimum eigenvalue.
+    The performance of AdaptVQE significantly depends on the minimization routine.
+    """
+
+    def __init__(
+        self,
+        solver: VQE,
+        excitation_pool: list[OperatorBase] = None,
```

Comment:
We will probably need to update the evolved operator anstaz too as we should be moving away from opflow. That may have to come later.

Reply:
I agree, this will have to come later. I actually went ahead and removed the `excitation_pool` argument for now. It is still a hard-coded requirement of this algorithm to use an `EvolvedOperatorAnsatz` so we can actually do what we have been doing so far, and extract the excitation pool from `ansatz.operators`. That way, using the `AdaptVQE` actually becomes a lot more user friendly (see the code-block example I added to the docstring and release note).